src/php/types/object.rs

```diff
@@ -157,6 +159,65 @@ impl ZendObject {
     fn mut_ptr(&self) -> *mut Self {
         (self as *const Self) as *mut Self
     }
+
+    /// Extracts some type from a Zend object.
+    ///
+    /// This is a wrapper function around `FromZendObject::extract()`.
+    pub fn extract<T: FromZendObject>(&self) -> Result<T> {
+        T::extract(self)
+    }
+}
+
+/// `FromZendObject` is implemented by types which can be extracted from a Zend object.
+///
+/// Normal usage is through the helper method `ZendObject::extract`:
+///
+/// ```rust,ignore
+/// let obj: ZendObject = ...;
+/// let repr: String = obj.extract();
+/// let props: HashMap = obj.extract();
+/// ```
+///
+/// Should be functionally equivalent to casting an object to another compatable type.
+pub trait FromZendObject: Sized {
+    /// Extracts `Self` from the source `ZendObject`.
+    fn extract(obj: &ZendObject) -> Result<Self>;
+}
```

Comment:
```suggestion
pub trait FromZendObject<'a>: Sized {
    /// Extracts `Self` from the source `ZendObject`.
    fn extract(obj: &'a ZendObject) -> Result<Self>;
}
```

Could we add a lifetime on the `ZendObject` reference? It will allow the trait to be implemented on types which will contain a reference to something inside the object (e.g. a property holding a string, returning a struct holding a `&str`).